### PR TITLE
Introduction to handles and handle pruning in compiler design

### LINKED ISSUE #5808: Introduction to handle and handle pruning in compiler design
#### Proposed article introduction
A handle is defined as a right-sentential-form that matches the right side of  a production and whose reduction to the left side non-terminal represents a single step along the reverse of a rightmost derivation.  

Handle pruning on the other hand is the process of identifying a handle and reducing it to the appropriate left hand side and it forms the basis of bottom up parsing in compiler design. Being that most compilers uses bottom up parsing strategy, handle pruning is considered key and very important in achieving parsing.

In this article I will be introducing a reader to the concept of handles through definition and discussing it at length for good understanding. I will then go ahead and introduce the reader to the concept of handle pruning and concluding with a formal definition of both. We will then go ahead to discuss using example the implementation of handle pruning as used in bottom up parsing.
A handle is defined as a right-sentential-form that matches the right side of  a production and whose reduction to the left side non-terminal represents a single step along the reverse of a rightmost derivation.  

#### Key takeaways 
1. The reader will be able to understand the concept of handle and handle pruning as used in bottom up parsing.
2. The will be able to identify a handle in a production and be able to reduce it to the start non-terminal during bottom up parsing.
3.  The reader will be able to implement handle pruning using code illustrations for bottom up parsing.

#### Article quality
This article will be made up of  illustrative content that is easy to understand even in the case of first time readers. It will also be exhaustive of all the concepts involved in handle and handle pruning based with a clearly defined introductory base. 

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by root5086**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?.

`This fixes #5808

### CODE REVIEW

File: content/articles/handle-and-handle-pruning/index.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+// Checking for valid productions
+checking();
+//Checks if the start symbol is on top of the stack
+if(stk[0] == 'S' && stk[1] == '\0')
+printf("Accept");
+else //do not allow
+printf("Rejection");
+}
+```
+### Conclusion
```
Comment (Jethro-magaji) [inline content/articles/handle-and-handle-pruning/index.md:243]: Please summarize what you have done in the article here.

File: content/articles/handle-and-handle-pruning/index.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+#### Introductory Background
+A handle is a concept used in the bottom-up parsing of compiler design. Bottom-up parsing is the method of syntax analysis in which sentences belonging to certain [context-free grammars](https://www.tutorialspoint.com/automata_theory/context_free_grammar_introduction.htm) whose role is to afford the construction of parse trees. This construction begins at the leaf nodes and proceeds to the root non-terminal hence the name bottom up which is sometimes called the shift-reduce parsing.
+
+Each symbol is moved onto the stack until there is a matching right-hand side nonterminal. This is followed by a reduction by the replacement of the production's right-hand side by its left-hand side. This process continues until eventually the string is reduced to the start non-terminal.
+
+The set of strings to be replaced at each reduction step is called a handle. 
+
+### Prerequisites
+The reader should have a prior understanding of the following concept:
```
Comment (Jethro-magaji) [inline content/articles/handle-and-handle-pruning/index.md:11]: Use 2nd person singular here "You". instead of "The reader".

**Review by Jethro-magaji — CHANGES_REQUESTED** [review]
Hi @root5086 
Well done for coming up with such an amazing article. However, there are a few corrections you need to make, please go over my reviews and make the corrections.
I suggest you run your article on tools like Grammarly and ProwritingAid to edit your work.
Thanks.

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by root5086** [comment]
@Jethro-magaji .....Corrections made.

**Review by Jethro-magaji — CHANGES_REQUESTED** [review]
Hi @root5086 
Well done for the edits. However, please make these minor changes, also explain each block of code below it.

**Comment by root5086** [comment]
Hi @Jethro-magaji ...the updates have been made...thanks

**Review by Jethro-magaji — APPROVED** [review]
Hi @root5086 !
Well done for the awesome article, please take note of the reviews next time you submit an article.
Thanks, overall it's a great article.

**Comment by hectorkambow** [comment]
jumping on this one

**Pull request merged**